### PR TITLE
Draft POC: Oauth2 redirect flow for authentication (instead of using popup)

### PR DESCRIPTION
**Pull request by headphonejames**
This is a proof of concept implementation for oauth2 authentication using a redirect flow from frontend -> backend API -> frontend, instead of using a secondary pop up to perform authentication. 

The high level changes look like this:
1. When redirect auth flow is enabled, accessing a resource that requires oauth2 permissions no longer triggers a window pop up for authentication. The front-end performs a redirect to the auth backend, sending the URI of the requesting page to the backend using a query string. 
2. After the back-end AuthConnector performs the authentication, it executes a redirect to the front end. Currently. the backstage auth and provider data is sent by cookie in the response.
3. The front-end SignInPage render checks the request for auth data. If auth cookie is present, the provider auth mechanism is triggered. 

This POC is missing many parts.  This POC illustrates the places in the code that must be modified to support this new flow with OAuth2, and well as validates this authentication flow works with github authentication. 

to do:
config integration for front end
introducing a separate auth handler for redirect
handling authentication failure
always trigging authentication modal dialog before sending request to auth backend informing user of redirect and potential loss of all app state data
fix login-failed msg in modal (race condition)
updating all providers to support redirect
testing serial authentication requests 
typescript correctness
SAML implementation
all the tests :)

open question:
It is possible access request headers in SignInPage component? If so, we could move from cookie to headers for transferring auth data.

The goal of this POC is to verify that this implementation approach makes sense before continuing. 

Thanks much!

James

### CODE REVIEW

**Comment by jamieklassen** [comment]
Forgive me if this is a silly question, but can you help describe the motivation for this change? Is it a "best practice" type of thing, or are there use cases where popups cause problems?

**Comment by dave-gantenbein** [comment]
> Forgive me if this is a silly question, but can you help describe the motivation for this change? Is it a "best practice" type of thing, or are there use cases where popups cause problems?

Yes, in some environments browsers are configured to block pop-ups. Ours is one, so this will remove a blocker for adoption for us. See here for further discussion: https://github.com/backstage/backstage/issues/9582

**Comment by headphonejames** [comment]
I updated the POC. Now configuration for auth popup propagated in both front and back end using new auth property "usePopup". The existing auth handler is used to manage both popup and redirect flow. 

Setting the correct "@backstage/core:SignInPage:provider" localstorage value before redirect is all that is required to trigger the session refresh using the correct provider after auth success redirect.

File: plugins/auth-backend/src/lib/oauth/OAuthAdapter.test.ts [hunk 1]
```diff
@@ -651,5 +718,60 @@ describe('OAuthAdapter', () => {
         sameSite: 'none',
       }),
     );
+    expect(mockResponse.redirect).toHaveBeenCalledTimes(0);
+  });
+
+  it('does not set the correct cookie configuration using origin from header when isPopupAuthenticationRequest is false', async () => {
```
Comment (dave-gantenbein) [inline plugins/auth-backend/src/lib/oauth/OAuthAdapter.test.ts:724]: Is there any way to extract or dry up some of the setup code in these tests?

**Comment by headphonejames** [comment]
Rebuilt this PR since my initial commit was not signed-off and I rebased from master after that. There are many commits that have conflicting rebase merges. I have submitted a PR that is passing all tests: https://github.com/backstage/backstage/pull/15841.

**Pull request closed without merging**